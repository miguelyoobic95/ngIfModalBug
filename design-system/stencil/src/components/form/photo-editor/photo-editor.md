---
name: PhotoEditor
category: Core Components
---

```photo-editor.html
    <yoo-photo-editor></yoo-photo-editor>
```

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|`text`|string|   |the text inside the label|

## CSS

|Type|Name|Description|
|---|---|---|
|Accent|`accent`|photo-editor accent|
|Dark|`dark`|photo-editor dark|
|Success|`success`|photo-editor success|
|Danger|`danger`|photo-editor danger|
|Info|`info`|photo-editor info|
|Warning|`warning`|photo-editor warning|
|Gradient Accent|`gradient-accent`|photo-editor gradient-accent|
|Gradient Dark|`gradient-dark`|photo-editor gradient-grey|
|Gradient Success|`gradient-success`|photo-editor gradient-success|
|Gradient Danger|`gradient-danger`|photo-editor gradient-danger|
|Gradient Info|`gradient-info`|photo-editor gradient-info|
|Gradient Warning|`gradient-warning`|photo-editor gradient-warning|

## Events
|Attr|Description|
|---|---|---|
|`yooXXXX`|small label|