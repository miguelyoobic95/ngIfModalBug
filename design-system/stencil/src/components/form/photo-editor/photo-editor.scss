@mixin mobile-styles() {
    top: 50%; /* position the top  edge of the element at the middle of the parent */
    left: 50%; /* position the left edge of the element at the middle of the parent */
    transform: translate(-50%, -50%);
    width: 100%;
}

:host {
    height: 100%;
    width: 100%;
    position: relative;
    background: cssvar(black);

    yoo-zoom {
        height: 100%;
        z-index: 500;
        background: cssvar(black);
        /deep/.zoom-container {
            ion-content {
                ion-scroll.scroll-ios {
                    background: none;
                }
            }
        }
    }
    .canvas {
        cursor: pointer;
        position: absolute;

        &.canvas-mobile {
            @include mobile-styles();
        }
    }

    .image {
        max-width: 100%;
        max-height: 100%;
        position: absolute;
        background: cssvar(light);

        .opacity-0 {
            opacity: 0;
        }

        &.image-mobile {
            @include mobile-styles();
        }
    }

    
}