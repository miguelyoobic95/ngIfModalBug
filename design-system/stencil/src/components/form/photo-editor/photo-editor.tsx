import { Component, Element, Prop, Event, EventEmitter } from '@stencil/core';
import { ICoreConfig } from '@shared/interfaces';
import { getProtocol, getElementDimensions } from '../../../utils/helpers';

@Component({
    tag: 'yoo-photo-editor',
    styleUrl: 'photo-editor.scss',
    scoped: true
})
export class YooPhotoEditorComponent {

    @Prop() readonly: boolean;
    @Prop() src: string;

    // @State() imageHasLoaded: boolean = false;

    @Event() imageClicked: EventEmitter<boolean>;
    @Element() host: HTMLStencilElement;

    imageTop: number = 0;
    imageLeft: number = 0;
    zoomComponent: HTMLYooZoomElement;
    canvasContext: any;
    private canvas: HTMLCanvasElement;
    private image: HTMLImageElement;
    // private colors: Array<string> = [Colors.assertive, Colors.balanced, Colors.positive, Colors.royal, Colors.energized, Colors.stable];
    private isLoading: boolean = false;
    private imageWidth: number = 0;
    private imageHeight: number = 0;

    private coreConfig: ICoreConfig = (window as any).coreConfigService;
    private isMobile: boolean;

    componentWillLoad() {
        // query selectors here seem to be giving errors
        this.coreConfig ? this.isMobile = this.coreConfig.isIonic() : this.isMobile = false;
    }

    componentDidLoad() {
        if (this.isMobile) {
            this.zoomComponent = this.host.querySelector('yoo-zoom');
        }
        this.canvas = this.host.querySelector('canvas');
        this.image = this.host.querySelector('.image');
        if (this.image && this.src) {
            this.isLoading = true;
            this.image.onload = () => this.onPhotoLoad();
            let src = this.src.replace('/upload/', '/upload/c_fill/').replace('http:', getProtocol());
            this.image.src = src;
        }
        // this.imageHasLoaded = true;
    }

    onPhotoLoad() {
        this.canvasContext = this.canvas.getContext('2d');
        this.centerPhoto();
    }

    onImageClicked() {
        this.imageClicked.emit(true);
    }

    centerPhoto() {
        this.imageWidth = this.image.width;
        this.imageHeight = this.image.height;
        const { height: hostHeight, width: hostWidth } = getElementDimensions(this.host);

        this.imageTop = (hostHeight - this.imageHeight) / 2;
        this.imageLeft = (hostWidth - this.imageWidth) / 2;
        // console.log('canvas before centering', this.canvas);
        // Match canvas to image
        this.canvas.width = this.imageWidth;
        this.canvas.height = this.imageHeight;
        // console.log('canvas after centering', this.canvas);
        // console.log('should have set canvas height', this.canvas, this.image);
    }

    setupListeners() {}

    renderWeb(): JSX.Element {
        return ([
            <img class="image" src={this.src}/>,
            <canvas class="canvas"></canvas>
        ]);
    }

    renderMobile() {
        return (
            <yoo-zoom>
                <img class={'image image-mobile ' + (this.isLoading ? 'opacity-0' : '')}/>
                <canvas class="canvas canvas-mobile" onClick={() => this.onImageClicked()}></canvas>
            </yoo-zoom>
        );
    }

    render(): JSX.Element {
        return (
            this.isMobile ? this.renderMobile() : this.renderWeb()
        );
    }
}
