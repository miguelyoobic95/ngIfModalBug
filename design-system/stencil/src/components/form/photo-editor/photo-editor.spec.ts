import { TestWindow } from '@stencil/core/testing';
import { YooPhotoEditorComponent } from './photo-editor';

describe('PhotoEditorComponent', () => {
    it('should build', () => {
        expect(new YooPhotoEditorComponent()).toBeTruthy();
    });

    describe('Rendering', () => {
        let element, window;
        beforeEach( async () => {
            window = new TestWindow();
            element = await window.load({
                components: [YooPhotoEditorComponent],
                html: `<yoo-photo-editor></yoo-photo-editor>`
            });
        });

        it ('Should render', async () => {
            expect(element).toMatchSnapshot();
        });

        it ('Should render with props', async () => {
            element.readonly = true;
            await window.flush();
            expect(element).toMatchSnapshot();
        });
    });

    describe('Events', () => {

        let emitEvent, window;
        beforeEach( async () => {
            window = new TestWindow();
            emitEvent = jest.fn();
        });
    });
});