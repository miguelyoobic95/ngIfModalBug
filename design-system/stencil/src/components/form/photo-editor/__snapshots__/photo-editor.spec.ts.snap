// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoEditorComponent Rendering Should render 1`] = `
<yoo-photo-editor
  class="hydrated"
  data-yoo-photo-editor-host=""
>
  
  <img
    class="image"
    data-yoo-photo-editor=""
  />
  <canvas
    class="canvas"
    data-yoo-photo-editor=""
  />
</yoo-photo-editor>
`;

exports[`PhotoEditorComponent Rendering Should render with props 1`] = `
<yoo-photo-editor
  class="hydrated"
  data-yoo-photo-editor-host=""
>
  
  <img
    class="image"
    data-yoo-photo-editor=""
  />
  <canvas
    class="canvas"
    data-yoo-photo-editor=""
  />
</yoo-photo-editor>
`;
